python/GafferOSLUI/OSLCodeUI.py

```diff
@@ -0,0 +1,320 @@
+##########################################################################
+#
+#  Copyright (c) 2016, Image Engine Design Inc. All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are
+#  met:
+#
+#      * Redistributions of source code must retain the above
+#        copyright notice, this list of conditions and the following
+#        disclaimer.
+#
+#      * Redistributions in binary form must reproduce the above
+#        copyright notice, this list of conditions and the following
+#        disclaimer in the documentation and/or other materials provided with
+#        the distribution.
+#
+#      * Neither the name of John Haddon nor the names of
+#        any other contributors to this software may be used to endorse or
+#        promote products derived from this software without specific prior
+#        written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+import functools
+
+import IECore
+
+import Gaffer
+import GafferUI
+import GafferOSL
+
+Gaffer.Metadata.registerNode(
+
+	GafferOSL.OSLCode,
+
+	"description",
+	"""
+	Allows arbitrary OSL shaders to be written directly within
+	Gaffer.
+	""",
+
+	"layout:customWidget:error:widgetType", "GafferOSLUI.OSLCodeUI._ErrorWidget",
+	"layout:customWidget:error:section", "Settings",
+	"layout:customWidget:error:index", -1,
+
+	plugs = {
+
+		"name" : [
+
+			"description", "Generated automatically - do not edit.",
+			"plugValueWidget:type", "",
+
+		],
+
+		"type" : [
+
+			"description", "Generated automatically - do not edit.",
+			"plugValueWidget:type", "",
+
+		],
+
+		"parameters" : [
+
+			"description",
+			"""
+			The inputs to the shader. Any number of inputs may be created
+			by adding child plugs. Supported plug types and the corresponding
+			OSL types are :
+
+			- FloatPlug (`float`)
+			- IntPlug (`int`)
+			- ColorPlug (`color`)
+			- V3fPlug (`vector`)
```

Comment:
It's an important distinction when it comes to using the `transform()` function, yes, and possibly some other cases that don't come to mind right now.

We do have [a ticket](https://github.com/ImageEngine/gaffer/issues/1819) for dealing with the distinction properly, ideally I'd like to deal with that separately.